### ISSUE TITLE
2.1.1 content negotiation failure — backcheck?

### ISSUE THREAD

**Issue by julianlam** [swicg/activitypub-html-discovery#23]
>  A more difficult failure mode to detect arises when the server does not support ActivityPub, but does support content negotiation for another JSON format. Such a server returns a 200 OK status code with a JSON object that does not use JSON-LD, or JSON-LD object that does not use the Activity Streams 2.0 vocabulary...

If one sends `Accept: application/activity+json` and received JSON back, you're right that you don't know whether it's JSON-LD. You could theoretically backcheck this by checking the received `Content-Type`, as I would _assume_ compliant AP implementations would send the appropriate header.

**Comment by evanp** [comment]
I think it's more of a situation when you use more general JSON types, like this:

```http
GET /resource HTTP/1.1
Host: mixed.example
Accept: application/activity+json, application/ld+json, application/json
```

I'll update the failure mode to be more clear that it's only when you accept non-AS2 JSON media types that compliant servers might not return AS2.

**Comment by evanp** [comment]
I think the work to be done here is to add a clarification about the Accept header that is sent and the JSON that is returned. Another example block with the GET request and Accept header should do the trick.

**Comment by trwnh** [comment]
I don't think requesting `application/json` should be used -- you should be requesting specifically activity+json or ld+json with the profile, and if those aren't available then the server can respond with a 405... although it might respond with a 200 and a default representation, if it "fails open" on content negotiation for unavailable types.

**Comment by TallTed** [comment]
> `Accept: application/activity+json, application/ld+json, application/json`

That's a poor `Accept:` header. There *must* be some preference order for these media types, and that should be included in the `Accept:` header (with `q=` notation, for perceived quality). The server can still ignore the header entirely, or deliver any of the listed media types (which might be due to calculated quality, which blends the *server's* quality ratings with the client's), or act entirely based on the client's `Accept:` header and deliver the client's highest quality rating.

**Comment by evanp** [comment]
> I don't think requesting `application/json` should be used -- you should be requesting specifically activity+json or ld+json with the profile

Some servers don't provide activity+json or ld+json -- for example, some static file servers; getting back application/json is better than nothing.

**Comment by evanp** [comment]
I updated this section with a little more information about how to detect this failure mode.

**Comment by trwnh** [comment]
@evanp wrote:

> some static file servers; getting back application/json is better than nothing.

Wouldn't static file servers be able to serve the JSON document as the default representation, without content negotiation?

**Comment by evanp** [comment]
> @evanp wrote:
> 
> > some static file servers; getting back application/json is better than nothing.
> 
> Wouldn't static file servers be able to serve the JSON document as the default representation, without content negotiation?

I don't understand this question or how it's relevant to the content negotiation failure section.

**Comment by trwnh** [comment]
Static file servers don't generally make use of content negotiation which is a dynamic feature. So a static JSON file will be served with Content-Type: application/json without necessitating Accept: application/json.

Also, if you receive application/json, then isn't that a failure? You can't be sure you got an AS2 document unless you do MIME type sniffing, which you generally shouldn't do, and you might even be instructed explicitly *not* to do this, via the nosniff directive: https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/X-Content-Type-Options#nosniff

**Comment by evanp** [comment]
> Static file servers don't generally make use of content negotiation which is a dynamic feature.

That's not true. If you have filename.ext1 and filename.ext2 in the same directory in a static file server, and there's a request for /filename, many static file servers will use content negotiation to determine which file to serve.

But that's neither here nor there. It's not what this issue is about.

What's important is this: it's helpful to developers to use application/json as a fallback. AS2 is valid JSON. I always do this and I recommend it to all ActivityPub developers.

But this document doesn't say to accept application/json or not. It says that if you accept application/json, you may get a JSON document that is not AS2. That's just a helpful note.

I don't see removing that warning as helpful to developers.
